lib/plug/telemetry.ex

```diff
@@ -0,0 +1,31 @@
+defmodule Plug.Telemetry do
+  @moduledoc """
+  A plug to instrument the pipeline with `:telemetry` events.
+  """
+
+  @behaviour Plug
+
+  @impl true
+  def init(_), do: []
+
+  @impl true
+  def call(conn, _opts) do
+    time = System.monotonic_time()
+    :telemetry.execute([:plug, :call, :start], %{}, %{conn: conn})
+
+    conn
+    |> Plug.Conn.put_private(:plug_telemetry_start_time, time)
+    |> Plug.Conn.register_before_send(&emit_stop_event/1)
+  end
+
+  defp emit_stop_event(conn) do
+    duration = System.monotonic_time() - conn.private[:plug_telemetry_start_time]
```

Comment:
Maybe these should specify the unit so that we can more clearly enforce the unit that the duration will have.
```suggestion
    duration = System.monotonic_time(:microsecond) - conn.private[:plug_telemetry_start_time]
```

Reply:
In the new relic agent, we capture everything in `native` time units and convert it when we know what unit we want, ie:

```elixir
duration = System.convert_time_unit(end_mono - start_mono, :native, :microsecond)
```

seems reasonable to do this for folks, it's not obvious

Reply:
I agree keeping it in native is best. Although maybe we should have a convenience in Telemetry.Metrics to convert from native to seconds.

Reply:
Or maybe make the unit configurable?